run dockerfile examples tests

alex1545 opened the pull request.

alex1545 commented:
/hold

nlopezgi commented on testing/e2e.sh:
```diff
@@ -542,6 +534,36 @@ function test_py_image_deps_as_layers() {
+function test_dockerfile_image_arbitrary() {
+  cd "${ROOT}/testing/examples/run_instruction_arbitrary"
+  clear_docker
+  bazel test --host_force_python=PY2 ...
```
add .bazelrc files to the projects with this flag

alex1545 replied:
Done.
There is only one "internal" project for all examples, so added one such file.

nlopezgi commented:
I'd rather we setup these new tests as GCB triggers if its not too much additional effort to do so.

alex1545 commented:
These tests will run as GCB jobs. Removed the tests from the `e2e.sh` script and added `cloudbuild.yaml` files for each of the examples. Single `cloudbuild.yaml` file won't be able to run the examples in parallel because it looks like bazel can be launched at most 4 times at a time.

These tests won't run for now, until this repo is configured to use GCB triggers for its PRs.

nlopezgi commented on contrib/dockerfile_build.bzl:
```diff
@@ -89,7 +89,7 @@ exports_files(["{}"])
             fail("Image build command failed: {} ({})".format(
                 build_result.stderr,
-                " ".join(command),
+                " ".join([str(e) for e in command]),
```
why is this needed, shouldn't all elements in command already be strings? If not, its probably better to make them strings as you are putting them in the command instead of converting them here

alex1545 replied:
Done.
Some of the elements were of type `path`.

nlopezgi commented on testing/examples/WORKSPACE:
```diff
@@ -16,9 +16,11 @@ workspace(name = "dockerfile_examples")
 load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive", "http_file")
 
-local_repository(
+http_archive(
```
Why the change from local_repository? You should leave it as local repo otherwise you would be testing a committed version of rules_docker not the changes you have in the PR

alex1545 replied:
Done.
Changed that when had trouble running targets (on GCB) in the "internal" bazel project within the rules_docker project. I figured it out now.

nlopezgi commented on testing/examples/basic/cloudbuild.yaml:
```diff
@@ -0,0 +1,29 @@
+
+steps:
+  # Confirm bazel and docker are installed.
+  - name: "gcr.io/asci-toolchain/nosla-ubuntu16_04-bazel-docker-gcloud"
```
use the bazel container (or do you need gcloud for some reason)?

alex1545 replied:
Done.
No, just need bazel and docker.

nlopezgi commented on testing/examples/basic/cloudbuild.yaml:
```diff
@@ -0,0 +1,29 @@
+  - name: "gcr.io/asci-toolchain/nosla-ubuntu16_04-bazel-docker-gcloud"
+    entrypoint: "sh"
+    args: ["-c", "cd testing/examples && bazel test //basic:all"]
+    waitFor: ["versions"]
```
use 'dir:' to set the directory to testing/examples (Example: https://github.com/bazelbuild/bazel-toolchains/blob/master/tests/rbe_repo/examples/rbe_custom.yaml#L8)
That way you dont need an entrypoint and the only args you should pass are 
 - test 
 - //basic:all

alex1545 replied:
Done.

nlopezgi commented on testing/examples/basic/cloudbuild.yaml:
```diff
@@ -0,0 +1,29 @@
+  # Confirm bazel and docker are installed.
+  - name: "gcr.io/asci-toolchain/nosla-ubuntu16_04-bazel-docker-gcloud"
+    entrypoint: "sh"
+    args: ["-c", "bazel version && docker version"]
```
No need to have a step just to check the bazel version and docker version

alex1545 replied:
Done.

nlopezgi commented on testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml:
```diff
@@ -0,0 +1,35 @@
+  # Test the run_instruction_apt_pkgs example.
+  - name: "gcr.io/asci-toolchain/nosla-ubuntu16_04-bazel-docker-gcloud"
+    entrypoint: "sh"
+    args: ["-c", "cd ../testing/examples && bazel --output_base=/workspace/apt_pkgs test //run_instruction_apt_pkgs:all"]
```
Why do you need to set the output base?

alex1545 replied:
Not a 100% sure, but this makes sure that Bazel generated files can be a accessed by multiple containers (even within the same steps). This is needed here and in the run_instruction_arbitrary example because these examples use the `download_pkgs`, `install_pkgs` and `container_run_and_commit` rules that launch additional containers.

nlopezgi replied:
So do you get an error if you dont set this?
What type of error?
If its because these rules need to launch additional containers, there should be some other way to avoid issues; e.g., by passing `/var/run/docker.sock:/var/run/docker.sock` in `volumes` (https://cloud.google.com/cloud-build/docs/build-config)

alex1545 replied:
Ass discussed separately, added a comment why this is needed.

nlopezgi commented on testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml:
```diff
@@ -0,0 +1,33 @@
+# Tests that verify the example installing apt packages using rules_docker
+# and Dockerfile.
+
+timeout: 21600s
```
Does this really need all that time to run?

alex1545 replied:
Done.
Reduced timeout to 30 mins.

nlopezgi commented on testing/examples/run_instruction_apt_pkgs/cloudbuild.yaml:
```diff
@@ -0,0 +1,33 @@
+# and Dockerfile.
+
+timeout: 21600s
+options:
```
why does this need the highcpu machine?

alex1545 replied:
I thought it was worth it as it bring the time of this test from almost 13 mins to ~5 mins.

nlopezgi replied:
cool, leave with the highcpu then but remove the timeout then!

alex1545 replied:
Done.

nlopezgi commented on testing/examples/run_instruction_arbitrary/cloudbuild.yaml:
```diff
@@ -0,0 +1,26 @@
+# Tests that verify the example manipulating arbitrary commands using
+# rules_docker and Dockerfile.
+
+timeout: 21600s
```
same q as above

alex1545 replied:
Timeout is reduced.
High CPU here brings the test time from almost 20 mins to just under 10 mins.

alex1545 commented:
/lgtm
